accepted/future-releases/variance/feature-specification.md

```diff
@@ -0,0 +1,373 @@
+}
+
+void main() {
+  D<int> d = D<num>();
+  d.x(24); // OK.
+  d.x = (int i) {}; // Dynamic error.
+}
+```
+
+*The class `D` inherits a setter with argument type `void Function(X)` even
+though it is an error to declare such a setter in `D`. It would be easy to
+prohibit invocations of that setter on an instance of type `D<...>`, but the
+invocation could then be performed using an upcast to `C`, so there is no
+real protection against executing such methods on that instance.*
+
+*Note that the subclass _can_ be written in such a way that the potential
+dynamic type error is eliminated, if it is possible to write a useful
+implementation with a safe signature:*
+
+```dart
+class SafeD<in X> extends C<void Function(X)> {
+  set x(void Function(Never) value) { 
+    if (value is void Function(X)) super.x = value;
+  }
+}
+```
+
+*Otherwise, the modifier `covariant` can be used to avoid the compile-time
+error for the member signature:*
+
+```dart
+class ExplicitlyUnsafeD<in X> extends C<void Function(X)> {
+  set x(covariant void Function(X) value) => super.x = value;
+}
+```
+
+*This makes it possible to declare method implementations with unsafe
+signatures, even in the case where the relevant type parameters of the
+enclosing class use sound variance.*
+
+
+### Type Inference
+
+During type inference, downwards resolution produces constraints on type
+variables with a variance modifier, rather than fixing them to a specific
+value in a partial solution. Upwards resolution will then include those
+constraints.
+
+Detailed rules will be specified in [inference.md].
+
+[inference.md]: https://github.com/dart-lang/language/blob/master/resources/type-system/inference.md
+
+
+## Dynamic Semantics
+
+This feature causes the dynamic semantics to change in only one way:
+The subtype relationship specified in the section on the static analysis
+is different from the subtype relationship without this feature, and the
+updated rules are used during run-time type tests and type checks.
+
+
+## Migration
+
+This proposal supports migration of code using dynamically checked
+covariance to code where some explicit variance modifiers are used, based
+on language versions.
+
+We use the phrase _legacy library_ to denote a library which is written in
+a language version that does not support sound variance.
+
+
+### Legacy libraries seen from a soundly variant library
+
+When a library _L_ with sound variance imports a legacy library _L2_, the
+declarations imported from _L2_ are seen in _L_ as if they had been
+declared in the language with sound variance.
+
+*In other words, source code in _L2_ is seen as having variance modifiers
+available, but it is simply not using them.*
+
+
+### Soundly variant libraries seen from a legacy library
+
+When a legacy library _L_ imports a library _L2_ with sound variance, the
+declarations imported from _L2_ are _legacy erased_. This means that all
+variance modifiers in type parameter declarations are ignored.
```

Comment:
Does this work out?   Doesn't this mean I can recreate your "twisted hierarchy" soundness bug by viewing a contravariant class from a legacy library?

Reply:
I changed this to say that the legacy library simply gets the declaration-site based subtype relationships whenever it refers to types that involve declarations using declaration-site variance. Surely that's not only simpler, it is also much more useful.

I haven't written anything about the management of breakage associated with corelib changes.